doc/reset_factory/ResetFactoryHLD.md

```diff
@@ -0,0 +1,210 @@
+
+This HLD document described the requirements, architecture details of Reset factory feature in switches Sonic OS based.
+
+###  1.3. <a name='DefinitionsAbbreviations'></a>Definitions/Abbreviations
+N/A
+
+
+###  1.4. <a name='Overview'></a>Overview
+
+A factory reset is a process that restores the switch to its original manufacturer settings. This can be useful if the device's configuration has become corrupt or if the user wishes to erase all personalized configurations.
+
+
+###  1.5. <a name='Requirements'></a>Requirements
+####  1.5.1. <a name='FunctionalRequirements'></a>Functional requirements
+1. It should support resetting configuration to factory default settings.
+
+2. It should clear logs, tech-support and reboot-cause files, users history files and home directories.
+
+3. It should reboot the system to load and apply the new configuration.
+
+4. It should stop sonic.target to ensure that files are not deleted while other daemons are utilizing them.
+
+5. it should allow users/application to create their own factory default configurations or
+extend the reset factory functionality.
+
+
+####  1.5.2. <a name='ConfigurationManagementRequirements'></a>Configuration and Management Requirements
+We will create a command line tool /usr/bin/reset-factory which should provide following functionality:
+
+1. default: reset configurations to factory default, clear all system logs and files.
+
+2. keep-basic: preserves basic configurations only after boot. Logs and files will be deleted users history files and home directories.
+
+3. keep-all-config: preserves all configurations after boot. Logs and files will be deleted.
+
+4. only-config: reset configurations to factory default. Logs and files will be preserved.
+
+    
+##  2 <a name='Design'></a>Design
+	
+###  2.1 <a name='High-LevelDesign'></a>High-Level Design
+![diagram](diagram.png)
+
+(flow description in the chapter below)
+
+"reset-factory" script will create new config_db.json if needed using "config-setup factory" command.<br/>
+In addition, it will clear logs, tech-support and reboot-cause files, users history files and home directories.<br/>
+
+Steps:
+
+1. Stops sonic.target: 
+
+systemctl stop sonic.target
+
+2. If not "keep-all-config":
+
+config-setup factory <None/keep-basic/only-config>
+
+3. If not ("only-config" or “keep-basic”):
+
+* Delete bash , vim and python history files under "/home".
+* Delete any non-dotfiles in the home directories.
+
+4. If not "only-config":
+* Delete tech-support files under "/var/dump/"
+* Delete all files under "/var/log"
+* Delete all files and symlinks under "/host/reboot-cause/"
+
+5. Print to /var/log/systemlog:
+New log file that won't be cleared in reset-factory.
+It will allow user to track operations such as "reset-factory".
+
+6. Reboot
+
+
+## 3 <a name='Functionality'></a>Functionality
+###  3.1 <a name='Reset-Factory'></a>Reset-Factory
+
+##### Default
+It will generate the default config_db.json file by running command:<br/>
```

Comment:
What is really factory reset means? How is different from brand new device vs the factory reset device? 
Ex: Use case - if RMA device or send device back to vendor wanted to delete all the data for security reasons?
1. Remove /etc/sonic files 
2. users & passwd
3. dockers overlay filesystem

Reply:
There are a few use cases:
1. System was tested in the vendor's lab prior to shipment to customer. before shipment any configuration should be deleted. but software installation should remain as is.
2. Customer got to configuration conflict and wishes to start from scratch.
3. Wiping information for security reason, before moving the system (including the case of shipping back to the vendor).

I would say that we want the system to **function** as a newly manufactured system.
This does not mean that the file system must be in the same state.
We want to achieve that by:
-  Resetting configurations to defaults (config_db.json)
-  Clear files that may potentially affect the functionality or user experience :
    This is very challenging as sudo user can change almost any file/directory in the system.
    We will try to cover the files that might be changed by the system.
    If the user choose to change the file system directly rather than using the CLI/official APIs, then the responsibility falls on him.
    
 1. Remove /etc/sonic files:
 I believe we can clear the /etc/sonic directory under the upperdir of overlayfs (e.g. /host/image*/rw/etc/sonic)
 but I need to test it in different scenarios first.
 
 2. users & passwd:
This will be handled by [User MGMT](https://github.com/sonic-net/SONiC/pull/1048) feature.
but since Reset Factory might be merged before User MGMT, we will handle this also.
We will delete non-default users and configure the default password for default users (e.g. admin )
 
 3. dockers overlay filesystem:
 As discussed in the HLD review meeting, this will be covered by removing all docker containers on system.
 
 I will update this HLD with the latest changes very soon,